### PR TITLE
External AuthZ C++ Data plane enablement

### PR DESCRIPTION
**Pull request by nickrmc83**
Description:
This change defines C++ grpc bindings for the external AuthZ interface. In the istio Security WG we're planning on using this interface to provide transparent authentication of requests. As this interface lies on the data plane we wish to implement it in C++ to maintain consistent and predictable performance

The changes allow a C++ implementation built using bazel to import Envoy as a workspace dependency and generate the C++ gRPC bindings directly without the need for complex import and generation scripts.

Risk Level:
Medium
Testing:
N/A
Docs Changes:
N/A
Release Notes:
N/A

### CODE REVIEW

**Comment by junr03** [comment]
@lizan could you take a look at this?

File: api/envoy/api/v2/core/BUILD [hunk 1]
```diff
@@ -35,7 +35,7 @@ api_proto_library_internal(
     name = "base",
     srcs = ["base.proto"],
     visibility = [
-        ":friends",
+        "//visibility:public",
```
Comment (lizan) [inline api/envoy/api/v2/core/BUILD]: why did you need to change this?

Comment (nickrmc83) [reply api/envoy/api/v2/core/BUILD]: Without the public visibility qualifier and given the bazel WORKSPACE extract shown below, the following error is encountered:
```
# Envoy API definitions

## We import envoy_api as a sub-repo of the mono-repo
git_repository(
    name = "envoy",
    branch = "data-plane",
    remote = "https://github.com/thales-e-security/envoy.git",
    verbose = True,
)

# Create @envoy_api virtual repository.
load("@envoy//bazel:api_repositories.bzl", "envoy_api_dependencies")

envoy_api_dependencies()

load("@envoy_api//bazel:repositories.bzl", "api_dependencies")

# api_dependencies() must be called first to import the correct versions of dependencies.
api_dependencies()
```

```
ERROR: xxx/.cache/bazel/_bazel_xxx/xxx/external/envoy_api/envoy/service/auth/v2/BUILD:5:1: in proto_library rule @envoy_api//envoy/service/auth/v2:attribute_context: target '@envoy_api//envoy/api/v2/core:base' is not visible from target '@envoy_api//envoy/service/auth/v2:attribute_context'. Check the visibility declaration of the former target if you think the dependency is legitimate
```

It appears https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/BUILD#L15 allows the external authz service definition to use types defined in base.proto. However, external implementers of https://github.com/envoyproxy/envoy/blob/master/api/envoy/service/auth/v2/external_auth.proto are not afforded the same rights.

The options I considered were:

- Make the contents of base.proto public
- Move the required types into another proto file and make that public.

The chose the former for simplicity as well as not wanting to cause significant rework for existing filters.

Comment (lizan) [reply api/envoy/api/v2/core/BUILD]: I cannot reproduce your error with bazel 0.27.0 with a project importing envoy like

```
local_repository(
    name = "envoy",
    path = "/Users/lizan/github/lizan/envoy/7459",
)


load("@envoy//bazel:api_repositories.bzl", "envoy_api_dependencies")

envoy_api_dependencies()

load("@envoy//bazel:repositories.bzl", "GO_VERSION", "envoy_dependencies")
load("@envoy//bazel:cc_configure.bzl", "cc_configure")

envoy_dependencies()

load("@rules_foreign_cc//:workspace_definitions.bzl", "rules_foreign_cc_dependencies")

rules_foreign_cc_dependencies()

cc_configure()

load("@io_bazel_rules_go//go:deps.bzl", "go_register_toolchains", "go_rules_dependencies")

go_rules_dependencies()

go_register_toolchains(go_version = GO_VERSION)
```

The code in /Users/lizan/github/lizan/envoy/7459 contains this PR and reverted this line.

```
bazel build @envoy_api//envoy/service/auth/v2:external_auth_service_cc_grpc
```
works well too.

The visibility restriction is to avoid accidental unindented dependencies to core, so this shouldn't be changed.

Comment (nickrmc83) [reply api/envoy/api/v2/core/BUILD]: Hmmm. With bazel 0.25.2 and 0.27.1 on Linux I get the original error even with the WORKSPACE snippet provided. I'll dig more tomorrow.

Comment (nickrmc83) [reply api/envoy/api/v2/core/BUILD]: @lizan you were right. I will revert to a visibility of `:friends`.

**Comment by lizan** [comment]
@nickrmc83 can you fix DCO by rebase or squash? otherwise LGTM.

/wait

**Comment by sungsk** [comment]
Gentle bump - 
This PR seems ready to be merged, but was closed due to staleness. Can we expect this to be merged soon?

**Comment by dio** [comment]
Let's reopen this. And hope @nickrmc83 could help on fixing the DCO.

File: api/envoy/service/auth/v2/BUILD [hunk 2]
```diff
@@ -26,3 +26,11 @@ api_proto_library_internal(
         "//envoy/type:http_status",
     ],
 )
+
+api_cc_grpc_library(
```
Comment (htuch) [inline api/envoy/service/auth/v2/BUILD]: Why can't we build this under the hood of `api_proto_library_internal`, so we don't need to do explicit declaration for each of these? I'd prefer that things "just work" rather than having to boiler plate each individual thing we want to make a service consumable outside of Envoy.

Comment (nickrmc83) [reply api/envoy/service/auth/v2/BUILD]: This is something I pondered. I opted for this route because I thought it was unlikely that most services would require CC protos and services. Is the preference to add to `api_proto_library` a `require_cc` option analogous to `require_py`?

Comment (htuch) [reply api/envoy/service/auth/v2/BUILD]: Yep, I think that would be a good idea (the additional option). Maybe call it `grpc_services` and make it a bool.

Comment (nickrmc83) [reply api/envoy/service/auth/v2/BUILD]: @htuch and @lizan see latest changes whereby `api_cc_grpc_library` and `api_go_grpc_library` are used by `api_proto_library` when `has_services` is set. I've refactored some of the services so they make use of this change. Some services that are Go-specific have remained the same so they can specialize calls to `api_go_grpc_library` as necessary.

File: api/bazel/api_build_system.bzl [hunk 3]
```diff
@@ -116,22 +130,31 @@ def api_proto_library(
...
         visibility = ["//visibility:public"],
     )
     py_export_suffixes = []
     if (require_py == 1):
-        api_py_proto_library(name, srcs, deps, has_services)
+        api_py_proto_library(name, srcs, deps)
         py_export_suffixes = ["_py", "_py_genproto"]
 
+    # Optionally define gRPC services
+    if (has_services == 1):
```
Comment (htuch) [inline api/bazel/api_build_system.bzl]: Nit: `if has_services:`

File: api/envoy/service/trace/v2/BUILD [hunk 4]
```diff
@@ -5,7 +5,7 @@ licenses(["notice"])  # Apache 2
 api_proto_library_internal(
     name = "trace_service",
     srcs = ["trace_service.proto"],
-    has_services = 1,
+    has_services = 0,  # A go-only service due to the external dependency @opencensus_proto//opencensus/proto/trace/v1:trace_proto_go
```
Comment (htuch) [inline api/envoy/service/trace/v2/BUILD]: What is a go-only service? Aren't all services reasonably implemented in all languages?

File: api/bazel/api_build_system.bzl [hunk 5]
```diff
@@ -74,6 +74,23 @@ def api_go_grpc_library(name, proto, deps = []):
...
+        deps = deps,
+        proto_only = False,
+        grpc_only = True,
+        visibility = ["//visibility:public"],
+    )
+
+def _ToCanonicalLabel(label):
+    # //my/app and //my/app:app are the same label. In places we mutate the incoming label adding different suffixes
+    # in order to generate multiple targets in a single rule. //my/app:app_grpc_cc.
+    # Skylark formatters and linters prefer the shorthand label whilst we need te latter.
```
Comment (nickrmc83) [inline api/bazel/api_build_system.bzl]: s/te/the

File: api/envoy/service/trace/v2/BUILD (hunk 4 above)
Comment (nickrmc83) [reply api/envoy/service/trace/v2/BUILD]: Go-only in the sense that it has external language-specific dependencies that have only been mapped for Go and not C++ or python or ...

I can update the comment to make that clearer and perhaps leave a `TODO` for good measure.

Comment (htuch) [reply api/envoy/service/trace/v2/BUILD]: Sounds good. LGTM modulo this bit. Thanks!

**Comment by nickrmc83** [comment]
/retest

Doesn't look like the coverage timeout should have anything to do with these changes

File: api/bazel/api_build_system.bzl [hunk 6]
```diff
@@ -74,6 +74,23 @@ def api_go_grpc_library(name, proto, deps = []):
...
     )
 
+def api_cc_grpc_library(name, proto, deps = []):
+    cc_grpc_library(
+        name = name,
+        srcs = [proto],
+        deps = deps,
+        proto_only = False,
+        grpc_only = True,
+        visibility = ["//visibility:public"],
```
Comment (htuch) [inline api/bazel/api_build_system.bzl]: Should we name this `_api_cc_grpc_library` (same with others in this file) to discourage use outside of the `.bzl`?

File: api/bazel/api_build_system.bzl [hunk 7]
```diff
@@ -98,6 +111,7 @@ def api_proto_library(
         has_services = 0,
         linkstatic = None,
         require_py = 1):
+    this = ":" + name
```
Comment (htuch) [inline api/bazel/api_build_system.bzl]: s/this/relative_name/

File: api/bazel/api_build_system.bzl (hunk 6 above)
Comment (nickrmc83) [reply api/bazel/api_build_system.bzl]: Done in 2740d3eb6924e5aa3105b456f9438441fe3be6d4

File: api/bazel/api_build_system.bzl (hunk 7 above)
Comment (nickrmc83) [reply api/bazel/api_build_system.bzl]: Done in 2740d3eb6924e5aa3105b456f9438441fe3be6d4

**Comment by htuch** [comment]
@nickrmc83 can you merge master and we can ship this?

**Comment by nickrmc83** [comment]
This branch has just been updated with the final merge from master and so is ready to go.

**Comment by nickrmc83** [comment]
Looks like `envoy-linux` and `envoy-linux (bazel release)` both failed due to a time with the same test: `//test/extensions/transport_sockets/tls:ssl_socket_test`. I don't think they are a result of these changes.

**Comment by htuch** [comment]
@nickrmc83 sorry to do this, but I'm going to merge https://github.com/envoyproxy/envoy/pull/8003 first, since it also mucks with `api_build_system.bzl` but is a much larger change, so we want to get it in ASAP now that it is LGTMed.

**Comment by htuch** [comment]
@nickrmc83 OK, #8003 is now merged; if you can merge master we can get this PR in as well, thanks again.

**Comment by nickrmc83** [comment]
@htuch this should be good to go.
